src/main/scala/devices/montgomery/Montgomery.scala

```diff
@@ -273,7 +273,7 @@ abstract class Montgomery(val params: MontgomeryParams, busWidthBytes: Int)
 
   val outputCounterEnable = RegInit(1.U(1.W)) // make sure the counter only run once
   val outputCounter = Counter(0 to block-1, (outputCounterEnable === 1.U) && (outStatus === 1.U) && (out.io.enq.fire), (control === 0.U))._1
-  outputCounterEnable := Mux((control === 2.U), Mux(outputCounter === (realInputBlock.asUInt-1.U), 0.U, outputCounterEnable), 1.U)
+  outputCounterEnable := Mux((control === 2.U), Mux(outputCounter >= (realInputBlock.asUInt-1.U) && (out.io.enq.fire), 0.U, outputCounterEnable), 1.U)
```

Comment:
There are circular dependencies.

`out.io.enq.valid` depends on `outputCounterEnable`, `outputCounterEnable` depends on `out.io.enq.fire` and internally `out.io.enq.fire` depends on `out.io.enq.valid`

`out.io.enq.valid` should not depend on the enable signal of the counter. You should always set `valid` to true as long as the data is valid (i.e. `out32(outputCounter)`) regardless of whether the counter is going to increase in the next cycle (i.e. the enable signal). If there is a transaction (i.e. `out.io.enq.fire`), you can then move on by enabling the counter.

Also, your `outputCounterEnable` is not true *enable* as it also depends on other conditions like  `(outStatus === 1.U) && (out.io.enq.fire)`. And, `(out.io.enq.fire)` is asserted twice for the enable signal, which is not meaningful.